### ISSUE TITLE
runAfter called before all data is updated

### ISSUE THREAD

**Issue by pduchesne** [vega/vega#1122]
In the following spec, i'm listening on click events to fill a selection of objects.

```javascript
{
    "data": [
        {"name": "source"},
        {
            "name": "selected",
            "on": [
                {"trigger": "clicked", "toggle": "clicked"}
            ]
        }
    ],
    "signals": [
        {
            "name": "clicked", "value": null,
            "on": [
                {
                    "events": "click",
                    "update": "datum",
                    "force": true
                }
            ]
        }
    ]
}
```

Then using the View API, I want to listen on this `clicked` signal to trigger code on selection change. I would expect this to work :
```
view.addSignalListener('clicked', function() {
    view.runAfter(function(view) {
        var selectedObjects = view.data("selected");
        console.warn("objects selected: "+selectedObjects.length);
    })
});
```
However the callback is invoked before the `selected` array is updated (although I use the `runAfter` method as documented).
It turns out that, for such a spec, the dataflow cycle triggered by a click ends before inserting the data into `selected` (thereby triggering my `runAfter` callback), and then starts a new dataflow cycle. This is suggested in the [runAsync documentation](https://vega.github.io/vega/docs/api/view/#view_runAsync).
So the only hackish solution that works AFAIK is relying on the promise of `runAsync`  : 
```
view.addSignalListener('clicked', function() {
    view.runAsync().then(function(view) {
        var selectedObjects = view.data("selected");
        console.warn("objects selected: "+selectedObjects.length);
    })
});
```
But this raises an error because you cannot trigger a new run within an signal handler.

It would be good to somehow have a `runafter` that behaves like the `runAsync` promise.

**Comment by jheer** [comment]
Thanks for filing the issue! Some Vega updates can not be completed in a single dataflow evaluation; trigger handling and autosize updates are two examples. Dataset modification in response to a trigger is handled via a `runAfter` call that re-invokes dataflow evaluation ([modify.js#L42](https://github.com/vega/vega-parser/blob/master/src/parsers/expression/modify.js#L42)). However, this call appears to be getting queued _after_ your signal listener's `runAfter` callback.

I will investigate further to try and make the behavior here more reasonable for View API users. In the meantime, you could try using a zero millisecond `setTimeout` call to delay your signal listener. That should ensure it runs after the dataflow evaluation(s).

**Comment by jheer** [comment]
Fixed in Vega 3.1. Vega now internally tracks a priority score for runAfter requests. Internally-triggered dataflow re-executions are handled first, such that the data state will be updated before API client callbacks are invoked.

**Comment by Casyfill** [comment]
I think I have the same issue:

```
var VIEW;

vegaEmbed('#vis', specs, opt)
	.then(function(result) {
		VIEW = result.view;
		VIEW.hover().run().runAfter(function(r){console.log(r.data('price'))})
	})
	.catch(console.warn);
```
does not return `price` data. even through it works through console, and I can access
console.log(VIEW._runtime.data.pis) (but not it's `output` neither `values.value`) in the same scope
